source/distributed/tria.cc

```diff
@@ -3506,7 +3506,10 @@ namespace parallel
             cell->clear_refine_flag ();
             cell->clear_coarsen_flag ();
           }
-
+      this->prepare_coarsening_and_refinement ();
+      // If flags come out after mesh smoothing on ghost and artificial cells,
+      // we should keep them because they are resulted from smoothing of locally
+      // owned active cells. So no cleaning is needed here.
```

Comment:
clean refinement flags on ghost cells before or after do make difference. @tjhei has proposed a very good example. Imagine we have a 3*3 initial grid on 9 processors, if we clean flags refinement on ghost cells before smoothing, and the smoothing flag `eliminate_refined_inner_islands` is set, then the central cell will never get refined.